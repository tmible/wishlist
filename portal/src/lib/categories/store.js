import { initCategories } from '$lib/categories/use-cases/init-categories.js';

/** @typedef {import('./domain.js').Category} Category */

/** @module Адаптер хранилища категорий */

/** @typedef {(value: any) => void} Subscriber */
/** @typedef {() => void} Unsubscriber */

/**
 * Значение хранилица
 * @type {Category[]}
 */
let value = [];

/**
 * Подписчики хранилища
 * @type {Map<Subscriber, Subscriber>}
 */
const subscriptions = new Map();

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * категорий
 * @type {object} Store
 * @property {(subscriber: Subscriber) => Unsubscriber} subscribe Подписка на значение хранилища
 * @property {() => Category[]} get Получение значения хранилища
 * @property {(newValue: Category[]) => void} set Установка значения хранилища
 * @property {(category: Category) => void} add Добавление категории в хранилище
 * @property {(category: Category) => void} update Обновление категории в хранилище
 * @property {(category: Category) => void} delete Удаление категории из хранилища
 */
export const categories = {
  subscribe: (subscriber) => {
    subscriber(value);
    subscriptions.set(subscriber, subscriber);

    if (subscriptions.size === 1) {
      initCategories();
    }

    return () => subscriptions.delete(subscriber);
  },

  get: () => value,

  set: (newValue) => {
    value = newValue;
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },

  add: (category) => {
    value.push(category);
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },

  update: (category) => {
    const categoryIndex = value.findIndex(({ id }) => id === category.id);
    if (categoryIndex === -1) {
      return;
    }
    value[categoryIndex] = category;
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },

  delete: (category) => {
    const categoryIndex = value.findIndex(({ id }) => id === category.id);
    if (categoryIndex === -1) {
      return;
    }
    value.splice(categoryIndex, 1);
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },
};
