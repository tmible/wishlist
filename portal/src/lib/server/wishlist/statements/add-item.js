import ListItemState from '@tmible/wishlist-common/constants/list-item-state';
import { inject } from '@tmible/wishlist-common/dependency-injector';
import { subscribe } from '@tmible/wishlist-common/event-bus';
import { Database } from '$lib/server/db/injection-tokens.js';
import { AddItem } from '../events.js';

/**
 * Создание SQL выражения для добавления элемента к списку желаний.
 * Подписка выполнения SQL выражения на соответствующее событие
 * @function initAddItemStatement
 * @returns {void}
 */
export const initAddItemStatement = () => {
  const statement = inject(Database).prepare(`
    INSERT INTO list (userid, name, description, state, "order", category_id, added_by)
    VALUES (?, ?, ?, ${ListItemState.FREE}, ?, ?, ?)
    RETURNING id
  `);
  subscribe(
    AddItem,
    (
      userid,
      {
        name,
        description,
        order,
        categoryId,
        addedBy = null,
      },
    ) => statement.get(userid, name, description, order, categoryId, addedBy),
  );
};
