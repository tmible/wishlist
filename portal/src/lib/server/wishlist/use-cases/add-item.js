import { inject } from '@tmible/wishlist-common/dependency-injector';
import { emit } from '@tmible/wishlist-common/event-bus';
import { RunTransaction } from '$lib/server/db/events.js';
import { IPCHub } from '$lib/server/ipc-hub/injection-tokens.js';
import { AddItem, GetItem, InsertDescriptionEntities } from '../events.js';

/** @typedef {import('$lib/wishlist/domain.js').OwnWishlistItem} OwnWishlistItem */
/** @typedef {import('$lib/wishlist/domain.js').OwnWishlistItemFormData} OwnWishlistItemFormData */

/** @module Сценарий добавления элемента к списку желаний */

/**
 * Добавление элемента
 * @function addItem
 * @param {number} userid Идентифкатор пользователя — владельца списка
 * @param {OwnWishlistItemFormData} item Добавляемый элемент
 * @returns {OwnWishlistItem['id']} Идентификатор добавленного элемента
 */
export const addItem = (userid, item) => {
  let addedItem;
  emit(
    RunTransaction,
    () => {
      const { id: itemId } = emit(AddItem, userid, item);
      emit(InsertDescriptionEntities, itemId, item.descriptionEntities);
      addedItem = emit(GetItem, itemId);
    },
  );
  inject(IPCHub).sendMessage(`update ${userid}`);
  return addedItem;
};
