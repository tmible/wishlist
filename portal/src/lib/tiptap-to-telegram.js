/** @module Перевод дерева текстового редактора в набор элементов разметки текста телеграма */

/** @typedef {import('@tiptap/core').JSONContent} JSONContent */
/** @typedef {import('$lib/components/telegram-entities/parser.svelte').Entity} Entity */
/**
 * @template ArrayType
 * @typedef
 * {ArrayType extends readonly (infer ElementType)[] ? ElementType : never}
 * ArrayElement<ArrayType extends readonly unknown[]>
 */
// eslint-disable-next-line no-secrets/no-secrets -- Просто длинное название
/**
 * Уплощёный узел дерева текстового редактора
 * @typedef {object} FlatChild
 * @property {string[]} marks Набор активных меток
 * //Будет работать только в TS
 * //@property {Pick<ArrayElement<Pick<JSONContent, 'marks'>>, 'type'>[]} marks Набор активных меток
 * @property {Pick<JSONContent, 'text'>} text Содержащийся в узле текст
 * @property {Pick<JSONContent, 'attrs'>} attrs Атрибуты для меток
 */

/**
 * Отображение названий меток текстового редактора в типы элементов
 * разметки текста телегарма Не упомянутые совпадают
 * @constant {Map<string, Pick<Entity, 'type'>>}
 */
const TIPTAP_TO_TELEGRAM_NAMES_MAP = new Map([
  [ 'strike', 'strikethrough' ],
  [ 'codeBlock', 'pre' ],
  [ 'link', 'url' ],
]);

/**
 * Рекурсивное уплощение дерева текстового редактора
 * @function flatContent
 * @param {JSONContent} content Текущий узел дерева
 * @param {string[]} [mixin] Примесь меток от родительских узлов дерева
 * @returns {FlatChild[]} Массив уплощённых потомков текущего узла
 */
const flatContent = (content, mixin = []) => content?.flatMap((child) => {
  if (child.type === 'text') {
    const url = child.marks?.find(({ type }) => type === 'text_link')?.attrs.href;
    return {
      marks: [ ...mixin, ...(child.marks?.map(({ type }) => type) ?? []) ],
      text: child.text,
      ...(url ? { attrs: { url } } : {}),
    };
  }

  if (child.type === 'codeBlock') {
    return {
      marks: [ child.type ],
      text: child.content[0].text ?? '',
      attrs: child.attrs,
    };
  }

  if (child.type === 'blockquote') {
    return flatContent(child.content, [ ...mixin, child.type ]);
  }

  const childContentFlat = flatContent(child.content, mixin) ?? [];
  if (child.content) {
    childContentFlat.at(-1).text += '\n';
  }
  return childContentFlat;
});

/**
 * Дополнительные проверки для приведения к соответствию формату Телеграма
 * @function preAndTextLinksMapper
 * @param {Entity} entity Элемент разметки текста
 * @returns {Entity} Обновлённый элемент разметки текста
 */
const preAndTextLinksMapper = (entity) => {
  if (entity.type === 'pre' && entity.language === null) {
    delete entity.language;
  }
  if (entity.type === 'text_link' && !entity.url.endsWith('/')) {
    entity.url += '/';
  }
  return entity;
};

/**
 * Перевод дерева текстового редактора в набор элементов разметки текста телеграма
 * 1. [Уплощение дерева]{@link flatContent};
 * 2. Перебор получившихся блоков:
 * 2.1. Создание элементов разметки текста на основе меток, закончившихся на начале текущего блока;
 * 2.2. Добавление к активным меткам меток, начавшихся в начале текущего блока;
 * 2.3. Добавление к тексту текста текущего блока;
 * 3. Удаление пробельных символов с конца текста;
 * 4. Создание элементов разметки текста на основе меток, закончившихся в конце текста;
 * @function tiptapToTelegram
 * @param {JSONContent} content Корень дерева текстового редактора
 * @returns {[ string, Entity[] ]} Текст и набор элементов его разметки
 */
export const tiptapToTelegram = (content) => {
  const contentFlat = flatContent(content);

  let text = '';
  const entities = [];
  let activeMarks = [];

  for (
    let i = 0, finishedMarks = [], [ block ] = contentFlat;
    i < contentFlat.length;
    ++i, text += block.text, block = contentFlat[i]
  ) {
    ({ activeMarks = [], finishedMarks = [] } = Object.groupBy(
      activeMarks,
      ({ type }) => (block.marks.includes(type) ? 'activeMarks' : 'finishedMarks'),
    ));

    entities.push(
      ...finishedMarks
        .map((finishedMark) => ({
          ...finishedMark,
          type: TIPTAP_TO_TELEGRAM_NAMES_MAP.get(finishedMark.type) ?? finishedMark.type,
          length: text.length - finishedMark.offset,
        }))
        .map((entity) => preAndTextLinksMapper(entity)),
    );

    activeMarks.push(
      ...block.marks
        .filter((mark) => !activeMarks.some(({ type }) => type === mark))
        .map((mark) => ({ ...block.attrs, offset: text.length, type: mark })),
    );
  }

  text = text.trimEnd();

  entities.push(
    ...activeMarks
      .map((activeMark) => ({
        ...activeMark,
        type: TIPTAP_TO_TELEGRAM_NAMES_MAP.get(activeMark.type) ?? activeMark.type,
        length: text.length - activeMark.offset,
      }))
      .map((entity) => preAndTextLinksMapper(entity)),
  );

  return [ text, entities ];
};
