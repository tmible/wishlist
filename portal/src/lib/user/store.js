import { DEFAULT_USER } from './default-user.const.js';

/** @typedef {import('./domain.js').User} User */

/** @module Адаптер хранилища пользователя */

/** @typedef {(value: any) => void} Subscriber */
/** @typedef {() => void} Unsubscriber */

/**
 * Значение хранилица
 * @type {User}
 */
let value = DEFAULT_USER;

/**
 * Подписчики хранилища
 * @type {Map<Subscriber, Subscriber>}
 */
const subscriptions = new Map();

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * пользователя
 * @type {object} Store
 * @property {(subscriber: Subscriber) => Unsubscriber} subscribe Подписка на значение хранилища
 * @property {() => User} get Получение значения хранилища
 * @property {(newValue: User) => void} set Установка значения хранилища
 */
export const user = {
  subscribe: (subscriber) => {
    subscriber(value);
    subscriptions.set(subscriber, subscriber);
    return () => subscriptions.delete(subscriber);
  },
  get: () => value,
  set: (newValue) => {
    value = newValue;
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },
};
