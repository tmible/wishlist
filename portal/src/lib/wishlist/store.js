/** @typedef {import('./domain.js').OwnWishlistItem} OwnWishlistItem */
/** @typedef {import('./use-cases/reorder-list.js').ReorderPatch} ReorderPatch */

/** @module Адаптер хранилища списка желаний */

/** @typedef {(value: any) => void} Subscriber */
/** @typedef {() => void} Unsubscriber */

/**
 * Значение хранилица
 * @type {OwnWishlistItem[] | null}
 */
let value = null;

/**
 * Подписчики хранилища
 * @type {Map<Subscriber, Subscriber>}
 */
const subscriptions = new Map();

/**
 * Уведомление подписчиков об изменении значения хранилища
 * @function notifySubscribers
 * @returns {void}
 */
const notifySubscribers = () => {
  for (const subscriber of subscriptions.values()) {
    subscriber(value);
  }
};

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * списка желаний
 * @type {object} Store
 * @property {(subscriber: Subscriber) => Unsubscriber} subscribe Подписка на значение хранилища
 * @property {() => OwnWishlistItem[]} get Получение значения хранилища
 * @property {(newValue: OwnWishlistItem[]) => void} set Установка значения хранилища
 * @property {(item: OwnWishlistItem) => void} add Добавление элемента списка в хранилище
 * @property {(item: OwnWishlistItem) => void} update Обновление элемента списка в хранилище
 * @property {(itemIds: OwnWishlistItem['id'][]) => void} delete
 *   Удаление элементов списка из хранилища
 * @property {(patch: ReorderPatch) => void} reorder Переупорядочивание списка в хранилище
 */
export const wishlist = {
  subscribe: (subscriber) => {
    subscriber(value);
    subscriptions.set(subscriber, subscriber);
    return () => subscriptions.delete(subscriber);
  },

  get: () => value,

  set: (newValue) => {
    value = newValue;
    notifySubscribers();
  },

  add: (item) => {
    if (!value) {
      value = [];
    }
    value.push(item);
    notifySubscribers();
  },

  update: (item) => {
    const itemIndex = value?.findIndex(({ id }) => id === item.id);
    if (itemIndex === -1) {
      return;
    }
    value[itemIndex] = item;
    notifySubscribers();
  },

  delete: (itemIds) => {
    let isAnyDeleted = false;

    for (const id of itemIds) {
      const itemIndex = value?.findIndex((item) => id === item.id);
      if (itemIndex === -1) {
        continue;
      }
      value.splice(itemIndex, 1);
      isAnyDeleted = true;
    }

    if (!isAnyDeleted) {
      return;
    }

    notifySubscribers();
  },

  reorder: (patch) => {
    if (!value) {
      return;
    }
    for (const item of value) {
      const reorder = patch.find(({ id }) => id === item.id);
      item.order = reorder?.order ?? item.order;
    }
    value.sort((a, b) => a.order - b.order);
    notifySubscribers();
  },
};
