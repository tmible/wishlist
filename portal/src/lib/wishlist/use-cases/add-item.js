import { inject } from '@tmible/wishlist-common/dependency-injector';
import { assignOrderToNewItem } from '../domain.js';
import { NetworkService, Store } from '../injection-tokens.js';

/** @typedef {import('../domain.js').OwnWishlistItem} OwnWishlistItem */

/** @module Сценарий добавления элемента к списку желаний */

/**
 * Запрос на добавление элемента, обновление хранилища
 * @function addItem
 * @param {Partial<OwnWishlistItem>} formData Добавляемый элемент
 * @returns {Promise<void>}
 * @async
 */
export const addItem = async (formData) => {
  const store = inject(Store);
  assignOrderToNewItem(store.get(), formData);
  const [ item, ok ] = await inject(NetworkService).addItem(formData);

  if (!ok) {
    return;
  }

  store.add(item);
};
