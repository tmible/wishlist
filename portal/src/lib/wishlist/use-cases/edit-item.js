import { inject } from '@tmible/wishlist-common/dependency-injector';
import { patchItem } from '../domain.js';
import { NetworkService, Store } from '../injection-tokens.js';

/** @typedef {import('../domain.js').OwnWishlistItem} OwnWishlistItem */

/** @module Сценарий изменения элемента списка желаний */

/**
 * Проверка наличия изменений в элементе списка, запрос на обновление, обновление хранилища
 * @function editItem
 * @param {OwnWishlistItem} item Изменяемый элемент списка
 * @param {Partial<OwnWishlistItem>} patch Объект с изменениями
 * @returns {Promise<void>}
 * @async
 */
export const editItem = async (item, patch) => {
  const [ isModified, itemPatched, patchFiltered ] = patchItem(item, patch);

  if (!isModified) {
    return;
  }

  const [ , ok ] = await inject(NetworkService).patchItem(item, patchFiltered);

  if (!ok) {
    return;
  }

  inject(Store).update(itemPatched);
};
