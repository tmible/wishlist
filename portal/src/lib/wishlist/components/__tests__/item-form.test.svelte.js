// @vitest-environment jsdom
import { cleanup, fireEvent, render, screen } from '@testing-library/svelte';
import { userEvent } from '@testing-library/user-event';
import { writable } from 'svelte/store';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { categories } from '$lib/categories/store.js';
import { tiptapToTelegram } from '$lib/tiptap-to-telegram.js';
import { addItem } from '../../use-cases/add-item.js';
import { editItem } from '../../use-cases/edit-item.js';
import ItemForm from '../item-form.svelte';

let categorySelectValue = null;

vi.mock('$lib/categories/store.js', () => ({ categories: writable([]) }));
vi.mock(
  '$lib/components/text-editor',
  async () => ({ default: await import('./mock.svelte').then((module) => module.default) }),
);
vi.mock('$lib/tiptap-to-telegram.js');
vi.mock('../../use-cases/add-item.js');
vi.mock('../../use-cases/edit-item.js');
vi.mock(
  '$lib/categories/select.svelte',
  () => ({ default: (_, props) => categorySelectValue = props.selectedCategoryId }),
);

describe('wishlist / components / item form', () => {
  afterEach(() => {
    categorySelectValue = null;
    vi.clearAllMocks();
    cleanup();
  });

  it('should be displayed with empty inputs', () => {
    const { container } = render(ItemForm);
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('should prefill inputs', () => {
    vi.mocked(categories).set([{ id: 'categoryId', name: 'categoryName' }]);
    const { container } = render(
      ItemForm,
      {
        values: {
          name: 'name',
          description: 'description',
          category: { id: 'categoryId', name: 'categoryName' },
        },
      },
    );
    vi.mocked(categories).set([]);
    expect(Array.from(container.querySelectorAll('input'), ({ value }) => value)).toMatchSnapshot();
    expect(categorySelectValue).toBe('categoryId');
  });

  it('should dispatch finish event on cancel button click', async () => {
    const onfinish = vi.fn();
    const user = userEvent.setup();
    render(ItemForm, { onfinish });
    await user.click(screen.getByText('Отмена', { selector: 'button' }));
    expect(onfinish).toHaveBeenCalled();
  });

  describe('on submit button click', () => {
    const props = $state({
      onfinish: vi.fn(),
    });

    let form;

    beforeEach(() => {
      vi.spyOn(
        Object,
        'fromEntries',
      ).mockReturnValue({
        categoryId: '',
        description: JSON.stringify({ content: 'content' }),
      });
      vi.mocked(tiptapToTelegram).mockReturnValue([]);
      const { baseElement } = render(ItemForm, props);
      form = baseElement.querySelector('form');
    });

    it('should call tiptapToTelegram', () => {
      fireEvent.submit(form);
      expect(tiptapToTelegram).toHaveBeenCalledWith('content');
    });

    it('should edit item', () => {
      props.values = { name: 'name' };
      props.item = { id: 'id' };
      fireEvent.submit(form);
      delete props.values;
      delete props.item;
      expect(vi.mocked(editItem)).toHaveBeenCalledWith({ id: 'id' }, expect.any(Object));
    });

    it('should add item', () => {
      fireEvent.submit(form);
      expect(vi.mocked(addItem)).toHaveBeenCalled();
    });

    it('should dispatch finish event', async () => {
      fireEvent.submit(form);
      await vi.waitFor(() => expect(props.onfinish).toHaveBeenCalled());
    });

    it('should set description', () => {
      tiptapToTelegram.mockReturnValueOnce([ 'description', [ 'entity 1', 'entity 2' ]]);
      fireEvent.submit(form);
      expect(
        vi.mocked(addItem),
      ).toHaveBeenCalledWith(
        expect.objectContaining({ description: 'description' }),
      );
    });

    it('should set description entities', () => {
      tiptapToTelegram.mockReturnValueOnce([ 'description', [ 'entity 1', 'entity 2' ]]);
      fireEvent.submit(form);
      expect(
        vi.mocked(addItem),
      ).toHaveBeenCalledWith(
        expect.objectContaining({
          descriptionEntities: [ 'entity 1', 'entity 2' ],
        }),
      );
    });

    describe('category', () => {
      describe('if there is no category id', () => {
        it('should set category to null', () => {
          fireEvent.submit(form);
          expect(
            vi.mocked(addItem),
          ).toHaveBeenCalledWith(
            expect.objectContaining({ category: null }),
          );
        });

        it('should delete categoryId', () => {
          fireEvent.submit(form);
          expect(
            vi.mocked(addItem),
          ).toHaveBeenCalledWith(
            expect.not.objectContaining({ categoryId: expect.anything() }),
          );
        });
      });

      describe('if there is category id', () => {
        beforeEach(() => {
          vi.mocked(
            Object.fromEntries,
          ).mockReturnValueOnce({
            categoryId: '123',
            description: JSON.stringify({ content: 'content' }),
          });
        });

        it('should set category from store if there is such', () => {
          vi.mocked(categories).set([{ id: 123, name: 'name 123' }]);
          fireEvent.submit(form);
          vi.mocked(categories).set([]);
          expect(
            vi.mocked(addItem),
          ).toHaveBeenCalledWith(
            expect.objectContaining({ category: { id: 123, name: 'name 123' } }),
          );
        });

        it('should set category to null if there is no such in store', () => {
          fireEvent.submit(form);
          expect(
            vi.mocked(addItem),
          ).toHaveBeenCalledWith(
            expect.objectContaining({ category: null }),
          );
        });

        it('should delete categoryId', () => {
          fireEvent.submit(form);
          expect(
            vi.mocked(addItem),
          ).toHaveBeenCalledWith(
            expect.not.objectContaining({ categoryId: expect.anything() }),
          );
        });
      });
    });
  });
});
