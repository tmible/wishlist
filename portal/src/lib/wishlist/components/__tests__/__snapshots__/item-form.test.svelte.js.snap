// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wishlist / components / item form > should be displayed with empty inputs 1`] = `"<form class="prose flex flex-col grow-1" method="POST"><div class="flex flex-col gap-2 grow-1 mb-6"><fieldset class="fieldset"><legend class="fieldset-legend">Название</legend> <input name="name" class="input w-full bg-base-200" required="" type="text" placeholder="Название"></fieldset> <fieldset class="fieldset"><legend class="fieldset-legend">Описание</legend> Mock<!----></fieldset> <fieldset class="fieldset"><legend class="fieldset-legend">Категория</legend> <!----></fieldset><!----></div> <div class="card-actions"><button class="btn btn-neutral w-full md:flex-1" type="button">Отмена</button> <button class="btn btn-primary w-full md:flex-1" type="submit">Добавить</button></div></form>"`;

exports[`wishlist / components / item form > should prefill inputs 1`] = `
[
  "name",
]
`;
